test/__init__.py

```diff
@@ -150,22 +150,27 @@ def add_file(self, file):  # type: (File) -> UserModule
     def url(self):  # type: () -> str
         return os.path.join('s3://', self.bucket, self.key)
 
-    def upload(self):  # type: () -> UserModule
-        with _files.tmpdir() as tmpdir:
-            tar_name = os.path.join(tmpdir, 'sourcedir.tar.gz')
-            with tarfile.open(tar_name, mode='w:gz') as tar:
-                for _file in self._files:
-                    name = os.path.join(tmpdir, _file.name)
-                    with open(name, 'w+') as f:
+    def create_tar(self, dir=os.path.dirname(os.path.realpath(__file__))):
+        tar_name = os.path.join(dir, 'sourcedir.tar.gz')
+        with tarfile.open(tar_name, mode='w:gz') as tar:
+            for _file in self._files:
+                name = os.path.join(dir, _file.name)
+                with open(name, 'w+') as f:
+
+                    if isinstance(_file.data, six.string_types):
+                        data = _file.data
+                    else:
+                        data = '\n'.join(_file.data)
 
-                        if isinstance(_file.data, six.string_types):
-                            data = _file.data
-                        else:
-                            data = '\n'.join(_file.data)
+                    f.write(data)
+                tar.add(name=name, arcname=_file.name)
+                os.remove(name)
```

Comment:
You dont need to delete the file here because everything is being written inside the tmp dir anyways.

Reply:
We need this because that tmpdir gets deleted when the function returns if you use a tmpdir for the tar file location. We use the contents of the tar file for the test `functional/test_download_and_impor.py::test_import_module_with_local_tar_via_download_and_extract` so in that use case, we are using the current directory (not tmpdir)